### PR TITLE
Auto SAC - Automated SAC (Selective Activation Checkpointing) Policy Construction and Wrapping 

### PR DESCRIPTION
**Pull request by sanketpurandare**
End-to-end workflow for automating Selective Activation Checkpointing to enable model training under memory constraints. Under the hood, this PR leverages several tools from `torch.distributed._tools`, namely `RuntimeEstimator`, `MemTracker`, `SACEstimator`, and `ILP for AutoSAC`. It first extracts the memory, runtime and activation checkpointing trade-off statistics. Using these it solves an ILP (Integer Linear Program) to determine the modules to AC and the memory budget for the selected modules. For each of the selected modules, it obtains a checkpointing policy (optimal or greedy) that determines whether to save or recompute the operators. Finally, the modules are wrapped using the policies.

Further, this PR adds the following:

1. Adds a Greedy Policy for determining whether to save or recompute an op using the ratio of (memory/runtime) as the greedy metric.
2. Adds support for estimating runtime by specifying a GPU type in `RuntimeEstimator` and `SACEstimator`.
3. Adds optimizer state memory, mandatorily saved memory, and upper bound for discarded memory in the SAC ILP. For FSDP composability, it estimates FSDP pre-fetch buffer, shared param, grad, and optimizer state memory.
4. Corrects the objective function of the SAC ILP to be bounded by total forward pass runtime.

*  #124688 
*  #134243 
*  #135208 
*  #137908 

End-to-end example with FSDP2:

```
from functools import partial
from typing import Callable, List, Tuple

import torch
from torch import nn, optim
from torch._subclasses.fake_tensor import FakeTensorMode
from torch.distributed._composable.fsdp import fully_shard
from torch.distributed._tensor import DeviceMesh
from torch.distributed._tools.auto_sac import (
    apply_auto_sac,
    get_auto_sac_policies,
    SACAlgorithm,
)
from torch.distributed._tools.fsdp2_mem_tracker import FSDPMemTracker
from torch.distributed._tools.mem_tracker import MemTracker
from torch.testing._internal.distributed._tensor.common_dtensor import (
    ModelArgs,
    Transformer,
)
from torch.testing._internal.distributed.fake_pg import FakeStore


def _init_model_input_optimizer(
    dev: torch.device,
) -> Tuple[Callable, List[nn.Module], List[optim.Optimizer], torch.Tensor]:
    bsz = 16
    model_args = ModelArgs(
        n_layers=4,
        n_heads=12,
        vocab_size=8192,
        max_seq_len=1024,
        dim=768,
        dropout_p=0.1,
    )
    with torch.device(dev):
        model = Transformer(model_args)
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-2, foreach=True)
    inp = torch.randint(
        0, model_args.vocab_size, (bsz, model_args.max_seq_len), device=dev
    )

    def train_step(
        models: List[nn.Module], optimizers: List[optim.Optimizer], inputs: torch.Tensor
    ) -> None:
        model = models[0]
        optimizer = optimizers[0]
        loss = model(inputs).sum()
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()

    return (train_step, [model], [optimizer], inp)


def _run_and_get_memtracker(
    train_step: Callable,
    models: List[torch.nn.Module],
    optimizers: List[torch.optim.Optimizer],
    inp: torch.Tensor,
) -> MemTracker:
    train_step(models, optimizers, inp)
    mem_tracker = MemTracker()
    mem_tracker.track_external(*models, *optimizers, inp)
    with mem_tracker:
        train_step(models, optimizers, inp)
    return mem_tracker


def _run_and_get_fsdp_memtracker(
    train_step: Callable,
    models: List[nn.Module],
    optimizers: List[optim.Optimizer],
    inp: torch.Tensor,
) -> FSDPMemTracker:
    fsdp_memtracker = FSDPMemTracker(models[0], optimizers[0])
    fsdp_memtracker.track_inputs((inp,))
    with fsdp_memtracker as fmt:
        for iter_idx in range(2):
            train_step(models, optimizers, inp)
            if iter_idx == 0:
                fmt.reset_mod_stats()
    if torch.distributed.group.WORLD:
        torch.distributed.destroy_process_group()
    return fsdp_memtracker


def _init_distributed(world_size: int) -> DeviceMesh:
    store = FakeStore()
    torch.distributed.init_process_group(
        "fake", rank=0, world_size=world_size, store=store
    )
    mesh = DeviceMesh("cuda", torch.arange(0, world_size))
    return mesh


def _apply_fsdp(models: List[nn.Module], optimizers: List[optim.Optimizer], mesh: DeviceMesh):
    model = models[0]
    fully_shard_fn = partial(fully_shard, mesh=mesh)
    for layer_id, transformer_block in enumerate(model.layers):
        reshard_after_forward = int(layer_id) < len(model.layers) - 1
        fully_shard_fn(transformer_block, reshard_after_forward=reshard_after_forward)
    fully_shard_fn(model)
    del optimizers[0]
    optimizers.append(
        torch.optim.Adam(models[0].parameters(), lr=1e-2, foreach=True)
    )


if __name__ == "__main__":
    fake_mode = FakeTensorMode()
    estimate_mode = "operator-level-cost-model"
    device = torch.device(type="cuda", index=torch.cuda.current_device())
    memory_budget = 3.0
    sac_algo = SACAlgorithm.OPTIMAL
    shard_degree = 4
    fsdp_units = {"Transformer.layers." + str(i) for i in range(4)} | {"Transformer"}
    mesh = _init_distributed(shard_degree)
    
    with fake_mode:
        train_step, models, optimizers, inp = _init_model_input_optimizer(device)

        mem_tracker = _run_and_get_memtracker(train_step, models, optimizers, inp)
        peak_mem_before = mem_tracker.get_tracker_snapshot("peak")[device]["Total"]

        auto_sac_result = get_auto_sac_policies(
            train_step,
            models,
            optimizers,
            inp,
            device,
            memory_budget=memory_budget,
            sac_algo=sac_algo,
            shard_degree=shard_degree,
            fsdp_units=fsdp_units,
            runtime_kwargs={
                "estimate_mode": estimate_mode,
            },
        )
        for model in models:
            apply_auto_sac(
                model, auto_sac_result.sac_policies, preserve_rng_state=False
            )
        
        _apply_fsdp(models, optimizers, mesh)

        fsdp_mem_tracker = _run_and_get_fsdp_memtracker(
            train_step, models, optimizers, inp
        )
        peak_mem_after = fsdp_mem_tracker.get_tracker_snapshot("peak")[device]["Total"]

    GiB = 2**30

    print(
        f"Peak Memory Before: {peak_mem_before / GiB:.2f} GiB\n"
        f"Memory Budget: {memory_budget} GiB\n"
        f"Auto-SAC Estimated Memory: {auto_sac_result.peak_mem / GiB} GiB\n"
        f"Peak Memory After: {peak_mem_after / GiB:.2f} GiB\n"
        f"Auto-SAC Decisions: {auto_sac_result.ac_decisions}"
    )
```
Output:
```
Peak Memory Before: 4.10 GiB
Memory Budget: 3.0 GiB
Auto-SAC Estimated Memory: 3.0 GiB
Peak Memory After: 2.60 GiB
Auto-SAC Decisions: {'Transformer.layers.0': 0.4282, 'Transformer.layers.1': 0.6138, 'Transformer.layers.2': 0.6138, 'Transformer.layers.3': 0.6138}
```

cc: @xuanzhang816 @weifengpy @gnadathur @lessw2020 
cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/_tools/sac_ilp.py [hunk 1]
```diff
@@ -103,6 +108,50 @@ def sac_milp(
             ):
                 prob += y[i] == 0
 
+        # Extract FQNs of leaf nodes from the graph
```
[the remark this answers is not in the document]
Comment (xuanzhang816) [reply torch/distributed/_tools/sac_ilp.py:111]: This is how I would code here (might have syntax errors, but the idea is this)
```suggestion
    node_to_param_size: Dict[str, int] = {
        node["fqn"]: node["param_per_module"] for node in graph.nodes
    }
    
    for fqn in graph["mod_order"]["fw_post_order"]:  # using post order so that children are visited first
        if fqn in fsdp_units:
            node = graph.name2node[fqn]
            j = node["index"]
            # its ancestor nodes will no longer need to take care of these parameters, so subtracting for them
            for i in range(node["index"]):
                if graph.ad_matrix[i][j]:
                    node_to_param_size[graph.nodes[i]["fqn"] -= node_to_param_size[fqn]
    
    # Find maximum parameter count among FSDP units
    max_fsdp_unit_memory = max(
        node_to_param_size[node["fqn"]] * (1 if node["fqn"] in fsdp_units else 0)
        for node in graph.nodes
    )
```

File: torch/distributed/_tools/sac_ilp.py [hunk 2]
```diff
@@ -103,6 +108,50 @@ def sac_milp(
             ):
```
Comment (sanketpurandare) [inline torch/distributed/_tools/sac_ilp.py:108]: Remove the AC units set part.

File: torch/distributed/_tools/sac_ilp.py [hunk 3]
```diff
@@ -103,6 +130,33 @@ def sac_milp(
...
+        }
+
+        for (
+            fqn
+        ) in graph.fw_post_order:  # using post order so that children are visited first
+            if fqn in fsdp_units:
+                node = graph.name2node[fqn]
+                j = node["index"]
+                # its ancestor nodes will no longer need to take care of these parameters, so subtracting for them
+                for i in range(node["index"]):
```
Comment (xuanzhang816) [inline torch/distributed/_tools/sac_ilp.py:144]: nit:
```suggestion
                for i in range(j):
```

File: torch/distributed/_tools/sac_ilp.py [hunk 4]
```diff
@@ -166,11 +229,24 @@ def sac_milp(
...
+        G_i = graph.nodes[i]["grad_per_module"] / MEM_MULTIPLIER
+        ACC_i = 0.0
+        for other_root_fqn in root_grad_mem:
+            if graph.name2node[other_root_fqn]["index"] > root_node["index"]:
+                ACC_i += root_grad_mem[other_root_fqn]
+        prob += m[i] == a[i] + 3 * max_fsdp_unit_memory[
+            root_node["fqn"]
+        ] + G_i + total_opt_memory + total_param_memory + ACC_i + (TG_i) / (
+            shard_degree if fsdp_units and root_node["fqn"] in fsdp_units else 1
+        )
```
Comment (xuanzhang816) [inline torch/distributed/_tools/sac_ilp.py:249]: Looks good to me. Thanks for making the modifications!

**Comment by weifengpy** [comment]
resolve conflicts?
